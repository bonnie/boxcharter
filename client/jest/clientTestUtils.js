import React from 'react'
import { Router } from 'react-router-dom'
import { Provider } from 'react-redux';
import { mount } from 'enzyme'
import { createMemoryHistory } from 'history'
// import { createStore } from 'redux';
// import App from '../src/app/App'

import { applyMiddleware, combineReducers, createStore } from 'redux';
import thunk from 'redux-thunk';
import { ReactWrapper } from 'enzyme';
// import { routerStateReducer } from 'redux-router';

// import * as generate from 'til-shared/generate'

// adapted from Kent Dodds's testing workshop repo
// https://github.com/kentcdodds/testing-workshop/blob/master/client/test/client-test-utils.js
function mountWithRouter(ui, {route = '/'} = {}) {
  const history = createMemoryHistory({initialEntries: [route]})
  const wrapper = mount(<Router history={history}>{ui}</Router>)
  return {
    history,
    wrapper,
    findNodes: findNodes.bind(null, wrapper),
    findNodeByTestId: findWrapperNodeByTestId.bind(null, wrapper),
  }
}

function sel(id) {
  return `[data-test="${id}"]`
}

function findNodes(wrapper, query) {
  // return wrapper.find(query).hostNodes()
  return wrapper.find(query)
}
/** end Kent dodds */

// this returns a new promise and is just a simple way to
// wait until the next tick so resolved promises chains will continue
function flushAllPromises() {
  return new Promise(resolve => setImmediate(resolve))
}

function findWrapperNodeByTestId(wrapper, id) {
  return findNodes(wrapper, sel(id))
}

// END: Kent Dodds helper functions

// adapted from 
// https://hackernoon.com/low-effort-high-value-integration-tests-in-redux-apps-d3a590bd9fd5
// function renderAppWithState(state) {
//   const store = createStore(state)
//   const wrapper = mount(
//     <Provider store={store}>
//       <App />
//     </Provider>
//   )
//   return [store, wrapper]
// }


// adapted from 
// https://medium.freecodecamp.org/real-integration-tests-with-react-redux-and-react-router-417125212638
/* Sets up basic variables to be used by integration tests
 * Params:
 *   reducers: should be an object with all the reducers your page uses
 *   initialRouterState: an optional object to set as the initial state for the router
 * Returns:
 *   an object with the following attributes:
 *     store: the reducer store which contains the main dispatcher and the state
 *     dispatchSpy: a jest spy function to be used on assertions of dispatch action calls
 */
// export function setupIntegrationTest(reducers, initialState = {}) {

//   /******* outdated, no longer necessary...? */
//   // // creating the router's reducer
//   // function routerReducer(state = initialRouterState, action) {
//   //   // override the initial state of the router so it can be used in test.
//   //   return routerStateReducer(state, action);
//   // }
//   /********** end: outdated, no longer necessary? */

//   // creating a jest mock function to serve as a dispatch spy for asserting dispatch actions if needed
//   const dispatchSpy = jest.fn(() => ({})); 
//   const reducerSpy = (state, action) => dispatchSpy(action);
//   // applying thunk middleware to the the store
//   const emptyStore = applyMiddleware(thunk)(createStore);
//   const combinedReducers = combineReducers({
//     reducerSpy,
//     router: routerReducer,
//     ...reducers,
//   });
//   // creating the store
//   const store = emptyStore(combinedReducers);

//   return { store, dispatchSpy };
// }

export {
  mountWithRouter,
  findWrapperNodeByTestId,
  sel,
  flushAllPromises,
  findNodes,
  // generate,
}
