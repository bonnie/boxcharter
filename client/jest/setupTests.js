// adapted from Kent Dodds's testing workshop
// https://github.com/kentcdodds/testing-workshop/blob/master/client/test/setup-tests.js

import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

Enzyme.configure({adapter: new Adapter()})

// here we set up a fake localStorage because jsdom doesn't support it
// https://github.com/tmpvar/jsdom/issues/1137
if (!window.localStorage) {
  window.localStorage = {}
  Object.assign(window.localStorage, {
    removeItem: function removeItem(key) {
      delete this[key]
    }.bind(window.localStorage),
    setItem: function setItem(key, val) {
      this[key] = String(val)
    }.bind(window.localStorage),
    getItem: function getItem(key) {
      return this[key]
    }.bind(window.localStorage),
  })
}


/* Udemy: 
import requestAnimationFrame from './tempPolyfills';
 
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
 
configure({ adapter: new Adapter(), disableLifecycleMethods: true });

*/