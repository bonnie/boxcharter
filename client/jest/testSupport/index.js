// adapted from https://github.com/jacek-rzrz/react-integration-tests-enzyme/blob/master/src/test_support/index.js

export * from './AppTestHelper';
export * from './enzymeHelpers';
export * from './actionsReducer';
export * from './mockApi';
export * from './asyncFlush';

