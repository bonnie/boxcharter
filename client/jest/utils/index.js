// from https://github.com/jacek-rzrz/react-integration-tests-enzyme/blob/master/src/utils/index.js

export * from './actionCreatorFactory';
export * from './locationChange';
export * from './checkErrors';
export * from './checkPropTypes';