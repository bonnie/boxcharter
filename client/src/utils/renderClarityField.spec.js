/*
 * Copyright (c) 2018 Bonnie Schulkin. All Rights Reserved.
 *
 * This file is part of BoxCharter.
 *
 * BoxCharter is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.
 *
 * BoxCharter is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with BoxCharter. If not, see <http://www.gnu.org/licenses/>.
 *
 */

/**
 * Tests for the renderClarityField component
 * @module
 * renderClarityField-spec
 */


import React from 'react'
import { shallow } from 'enzyme'
import '../../jest/setupTests'
import renderClarityField from './renderClarityField'

// const { input, type, label, required }  = props
// const { touched, error, warning } = props.meta

describe('renderClarityField', () => {
  test('renders a text field correctly', () => {
    const props = {
      type: 'text',
      label: 'Test',
      name: 'test',
      meta: {},
      input: {},
    }
    const renderedField = shallow(renderClarityField(props))
    expect(renderedField).toMatchSnapshot()
  })
  test('renders a required field correctly', () => {
    const props = {
      type: 'text',
      label: 'Test',
      name: 'test',
      required: true,
      meta: {},
      input: {},
    }
    const renderedField = shallow(renderClarityField(props))
    expect(renderedField).toMatchSnapshot()
  })
  test('renders an invalid field correctly', () => {
    const props = {
      type: 'text',
      label: 'Test',
      name: 'test',
      required: true,
      meta: { touched: true, error: 'bad field' },
      input: {},
    }
    const renderedField = shallow(renderClarityField(props))
    expect(renderedField).toMatchSnapshot()
  })
})