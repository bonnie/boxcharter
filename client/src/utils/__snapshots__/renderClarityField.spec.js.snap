// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderClarityField renders a required field correctly 1`] = `
<fieldset
  className="form-group"
>
  <label
    className="required"
  >
    Test
  </label>
  <label
    className="tooltip tooltip-validation "
  >
    <input
      className="form-control"
      type="text"
    />
    <span
      className="tooltip-content"
    />
  </label>
</fieldset>
`;

exports[`renderClarityField renders a text field correctly 1`] = `
<fieldset
  className="form-group"
>
  <label
    className=""
  >
    Test
  </label>
  <label
    className="tooltip tooltip-validation "
  >
    <input
      className="form-control"
      type="text"
    />
    <span
      className="tooltip-content"
    />
  </label>
</fieldset>
`;

exports[`renderClarityField renders an invalid field correctly 1`] = `
<fieldset
  className="form-group"
>
  <label
    className="required"
  >
    Test
  </label>
  <label
    className="tooltip tooltip-validation invalid"
  >
    <input
      className="form-control"
      type="text"
    />
    <span
      className="tooltip-content"
    >
      bad field
    </span>
  </label>
</fieldset>
`;
